src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlLanguageServerProtocol.cs

```diff
@@ -23,25 +25,42 @@ namespace Microsoft.VisualStudio.LanguageServices.Xaml.LanguageServer
     internal sealed class XamlLanguageServerProtocol : AbstractRequestHandlerProvider
     {
         private readonly XamlProjectService _projectService;
+        private readonly IXamlLanguageServerFeedbackService? _feedbackService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public XamlLanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, XamlProjectService projectService)
+        public XamlLanguageServerProtocol(
+            [ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, 
+            XamlProjectService projectService,
+            [Import(AllowDefault = true)] IXamlLanguageServerFeedbackService? feedbackService)
             : base(requestHandlers, languageName: StringConstants.XamlLanguageName)
         {
             _projectService = projectService;
+            _feedbackService = feedbackService;
         }
 
-        protected override Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(RequestExecutionQueue queue, RequestType request, ClientCapabilities clientCapabilities, string? clientName, string methodName, bool mutatesSolutionState, IRequestHandler<RequestType, ResponseType> handler, CancellationToken cancellationToken)
+        protected override async Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(RequestExecutionQueue queue, RequestType request, ClientCapabilities clientCapabilities, string? clientName, string methodName, bool mutatesSolutionState, IRequestHandler<RequestType, ResponseType> handler, CancellationToken cancellationToken)
         {
             var textDocument = handler.GetTextDocumentIdentifier(request);
 
+            DocumentId? documentId = null;
             if (textDocument is { Uri: { IsAbsoluteUri: true } documentUri })
             {
-                _projectService.TrackOpenDocument(documentUri.LocalPath);
+                documentId = _projectService.TrackOpenDocument(documentUri.LocalPath);
             }
 
-            return base.ExecuteRequestAsync(queue, request, clientCapabilities, clientName, methodName, mutatesSolutionState, handler, cancellationToken);
+            using (var requestScope = _feedbackService?.CreateRequestScope(documentId, methodName))
+            {
+                try
+                {
+                    return await base.ExecuteRequestAsync(queue, request, clientCapabilities, clientName, methodName, mutatesSolutionState, handler, cancellationToken).ConfigureAwait(false);
+                }
+                catch(Exception e) when (e is not OperationCanceledException)
+                {
+                    requestScope?.RecordFailure(e);
+                    throw;
```

Comment:
❔ Why isn't this using roslyn's standard non-fatal error reporting strategy?

/cc @ryzngard

Reply:
Here's what I expected to see:

```suggestion
                catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken))
                {
                    throw ExceptionUtilities.Unreachable;
```

Reply:
This is just to tell XLS a failure happened inside this RequestScope. It is not reporting any fault/watson for this one.

---
In reply to: [563985150](https://github.com/dotnet/roslyn/pull/50754#discussion_r563985150) [](ancestors = 563985150)

Reply:
/cc @dibarbet 

Here's what we are trying to achieve. Each time an LSP request throws we want to mark our request scope as failed. What's the best way to handle that (because the suggestion doesn't)?

Reply:
I was mostly going based on this:

> ... will aggregate the data and send telemetry ...

I'm not familiar with the LSP events themselves, but I wanted to make sure if the goal was to just get error telemetry we weren't making it more complicated than necessary. @dibarbet and @ryzngard would better know if any change to the implementation is needed.

Reply:
> Each time an LSP request throws we want to mark our request scope as failed

If you're not trying to record fault data here then `FatalError.ReportAndPropagateUnlessCanceled` should not be used. It looks like you're rethrowing though, which could mean that failure is going to be caught and reported elsewhere. 

What will you be doing with the knowledge that a RequestScope has failed? Will that be reported somehow?

Reply:
What does `RecordFailure` do?

If you want loghub logging, it should be possible as part of the request context - https://github.com/dotnet/roslyn/blob/master-vs-deps/src/Features/LanguageServer/Protocol/Handler/RequestContext.cs#L230

For error telemetry Sam's method should work.

Otherwise, if it is doing something specific to Xaml then this seems OK

Reply:
So the RecordFailure will inform XLS that this RequestScope failed. This could give us information like for a particular LSP method, what is the success rate of the requests - success count/total request count for example. We aggregate the counts on our end as we don't want to report events for every single requrests.

If the RecordFailure(Exception) is causing all the concerns, do you think if we switch to just record a flag, like HasException = true, will make it better?

Reply:
I'm not familiar with the specific use case and where telemetry would be helpful, but as a general trend we almost always record faults at point of failure. If that information is not useful to you or not actionable, this seems fine. Just add a comment explaining that aggregate data for failure rates is useful but individual callstacks are not.

Reply:
Sorry for the confusions. Also when debugging I see exceptions from ExecuteRequestAsync in general are reported inside RequestExecutionQueue already via ReportNonFatalErrorAsync, so I guess we should be fine and won't miss any important information for the exceptions.
https://github.com/dotnet/roslyn/blob/cffe3c1171f85891ac4ba873a7642d466bb8c706/src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs#L199

I will add comments that on the RecordFailure to make it more clear.

Reply:
@LinglingTong that reminds me of a possible issue - I don't think the exception will bubble up here, we intentionally catch exceptions for non-mutating operations so that we do not shutdown the server.  So I'm not certain you will ever get anything here, can you check?

Reply:
They should still bubble up to here because this is actually waiting on a TaskCompletionSource, and we do pass the exceptions back (see https://github.com/dotnet/roslyn/blob/cffe3c1171f85891ac4ba873a7642d466bb8c706/src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs#L154 ) we just don't let them bubble up to the execution queue process.

Reply:
Yeah I can get the exception if there is one.